Desktop: Add support for external codemirror plugins

CalebJohn opened the pull request:
Not much to show yet, I've mostly just copied the markdownIt support. I will be making some changes once I figure out how I want the interface to look.

laurent22 commented on ReactNativeClient/lib/services/plugins/utils/contentScriptsToCodeMirror.ts:
```diff
@@ -0,0 +1,28 @@
+import { PluginStates } from '../reducer';
+import { ExtraRendererRule } from 'lib/joplin-renderer/MdToHtml';
+import { ContentScriptType } from '../api/types';
+
+export default function contentScriptsToCodeMirror(plugins:PluginStates):ExtraRendererRule[] {
+	const output:ExtraRendererRule[] = [];
+
+	for (const pluginId in plugins) {
+		const plugin = plugins[pluginId];
+		const contentScripts = plugin.contentScripts[ContentScriptType.CodeMirrorPlugin];
+		if (!contentScripts) continue;
+
+		for (const contentScript of contentScripts) {
+			const module = require(contentScript.path);
+			if (!module.default || typeof module.default !== 'function') throw new Error(`Content script must export a function under the "default" key: Plugin: ${pluginId}: Script: ${contentScript.id}`);
+
+			const loadedModule = module.default({});
+			if (!loadedModule.plugin) throw new Error(`Content script must export a "plugin" key: Plugin: ${pluginId}: Script: ${contentScript.id}`);
+
+			output.push({
+				id: contentScript.id,
+				module: loadedModule,
+			});
+		}
+	}
+
+	return output;
+}
```
It's good that the logic is pretty much the same as for loading markdown-it plugins. But in that case, maybe we should refactor the existing contentScriptsToRendererRules so that it's more generic and use it for both CodeMirror and MarkdownIt? From the look of it, it will only need to take an extra "ContentScriptType" parameter.

CalebJohn replied:
That's what I'm thinking I'll do, I started like this in case I decided to make drastic changes, this was just a good starting point. I don't imagine I will make any sweeping changes because keeping the interface the same between content scripts is pretty valuable.

laurent22 commented on ElectronClient/gui/NoteEditor/NoteBody/CodeMirror/utils/useExternalPlugins.ts:
```diff
@@ -0,0 +1,15 @@
+		contentScript.module.plugin(CodeMirror);
+	}
+	// TODO: This should only be run on started, currently it is run on every key stroke
+	console.log('Loading CodeMirror Plugins');
```
I guess just wrapping it in useEffect would do the job. The "plugins" state variable is stable and won't change often.

CalebJohn replied:
Good call, nice and simple.

CalebJohn commented:
@laurent22 I've made a bit of progress on this, so the user can load codemirror plugins but I'm a bit stuck on loading assets (like css). I figured that for the markdownIt rules there must be a way to do this and I'd like to follow the same method. I was hoping you could give me a hint about where to look to understand that. After I get css loading this pull should be good to go.

laurent22 commented:
@CalebJohn, for the assets, the first step would be for the plugin to declare what they need. For this, have a look for example at mermaid.ts as you could use a similar structure:

```typescript
export default {

	assets: function() {
		return [
			{ name: 'mermaid.min.js' },
			{ name: 'mermaid_render.js' },
			{
				inline: true,
				text: '.mermaid { background-color: white; width: 640px; }',
				mime: 'text/css',
			},
		];
	},

	plugin: function(markdownIt:any) {
		// ...
	},
};
```

So each asset is basically:

```typescript
name: string, // Path to file
inline?: boolean, // Defaults to false
mime?: string, // Auto-detected from file extension when possible
text?: string, // For inline CSS or JS only
```

Then wherever you load this content script, you'll have the plugin under the `plugin` key, and the assets under the `assets` key.

To load them, it would depend on your CodeMirror component. Perhaps you can simply loop on the assets and create `script`, `link` or `style` element depending on the assets, as is done in note-viewer/index.html (`addPluginAssets` function)?

CalebJohn commented:
@laurent22 Thank you for the detailed information, that's exactly what I needed. 

This is now ready to merge, I only added support for adding css assets because js assets shouldn't be necessary. 

I added a new parameter `assetPath` when running `loadContentScript` to make it easier to attach css files. I'm not sure that this was the best way to handle it, so please let me know if there is something better.

I also noticed while working on this that loading assets for external markdown-it plugins probably won't work because they seem to make the assumption that they'll be in the pluginAssets directory. I didn't want to make changes to that here, but I can create another pull if you'd like.

laurent22 commented on ReactNativeClient/lib/joplin-renderer/MdToHtml.ts:
```diff
@@ -1,6 +1,7 @@
 import InMemoryCache from 'lib/InMemoryCache';
 import noteStyle from './noteStyle';
 import { fileExtension } from './pathUtils';
+import { ExtraContentScript } from 'lib/services/plugins/utils/loadContentScripts';
```
The renderer shouldn't have a dependency to the lib as it could potentially be used separately (that has become clearer with the Lerna update). Perhaps you could simply leave the ExtraRendererRule interface in this file since it's compatible with ExtraContentScript.

CalebJohn replied:
I'll keep that in mind going forward, thanks. This is fixed here.

laurent22 commented on ElectronClient/gui/NoteEditor/NoteBody/CodeMirror/utils/useExternalPlugins.ts:
```diff
@@ -0,0 +1,69 @@
+		for (const contentScript of contentScripts) {
+			const mod = contentScript.module;
+			for (const addon of mod.addons) {
+				require(`codemirror/addon/${addon}`);
```
If this is an external plugin, shouldn't it be at the plugin level that we require whatever is needed. What if someone create a brand new plugin that's not in codemirror/addon, I guess it wouldn't work in this case?

laurent22 replied:
In fact, it might help to add the documentation to the `ContentScriptType` interface, like it's done for the markdown-it. As knowing how it should work will help discuss the implementation details.

CalebJohn replied:
These are specifically for the [codemirror addons](https://codemirror.net/doc/manual.html#addons). Codemirror has an odd way of loading plugins that means we can't load them through the contrentScript plugin. Basically when you require a CodeMirror plugin it modifies the state of the already imported codemirror instance. All other requires can be done at the plugin level, so I think this separation makes sense.

I'll get on that documentation soon and hopefully that will help things makes sense.

laurent22 replied:
Ok that makes sense. Is there also a concept of external addon in CodeMirror (one that wouldn't already be packaged with the distribution)?

CalebJohn replied:
They exist, but I don't know of many. From what I have seen they usually don't use the special loading system that the codemirror built-in addons use. The system that we use internally of calling a function with `CodeMirror` as an argument seems to be normal. Our content scripts work exactly like this and pass `CodeMirror` to the function. This would allow plugin developers to require their own external plugin and pass `CodeMirror` on to it.

laurent22 commented:
> The system that we use internally of calling a function with CodeMirror as an argument seems to be normal. Our content scripts work exactly like this and pass CodeMirror to the function. This would allow plugin developers to require their own external plugin and pass CodeMirror on to it.

Yes that's a good idea, a simple function where we pass the CodeMirror instance and then the developper can do whatever they want with it. So how about we define a CodeMirror content script to be like?

```javascript
module.exports = {
    default: function(context) {
        return {
            plugin: function(codeMirror) {
                // ...
            },
            assets: {
                // ...
            },
        }
    }
}
```

We would keep your asset loading logic, and for any JavaScript logic, the developer can do as they want, adding modules, new modes, etc. using the CodeMirror instance. Do you think that would work?

CalebJohn commented:
@laurent22 
That's how this pull request implements it

```javascript
function plugin(CodeMirror) {
	CodeMirror.commands.printKey = function(cm) {
		console.log("Pressed Ctrl-U!");
	}
	CodeMirror.keyMap.basic["Ctrl-U"] = "printKey"
}

module.exports = {
	default: function(_context) { 
		return {
			plugin: plugin,
			addons: ['search/match-highlighter'],
			codeMirrorOptions: {'highlightSelectionMatches': true},
			assets: function() {
				return [
					{ name:'./selection.css' }
					// { inline: true,
						// text: '.cm-matchhighlight {	background-color: lightgreen;}',
						// mime: 'text/css',
					// }
				];
			},
		}
	},
}
```

Above is the plugin that I've been using for testing, it implements a new codemirror command and binds it to Ctrl-U.
It also enables the codemirror match highlighter addon and provides css for it. This css can be inline, or a separate file,

laurent22 reviewed (commented):
Thanks for the explanation, I had missed how it was working from looking at the code, but your implementation makes sense. I've added a few comments to the code but other than that it's almost ready to merge.

CalebJohn commented on packages/app-desktop/gui/NoteEditor/NoteBody/CodeMirror/utils/useExternalPlugins.ts:
```diff
@@ -0,0 +1,69 @@
+					if (asset.inline) {
+						cssStrings.push(asset.text);
+					} else {
+						addScript(resolve(contentScript.assetPath, asset.name), mod.id);
```
It should work on windows, here is an excerpt from the [path module documentation](https://nodejs.org/api/path.html#path_windows_vs_posix)
> On Windows, both the forward slash (/) and backward slash (\) are accepted as path segment separators; however, the path methods only add backward slashes (\).

Anyways I have changed I replaced it with `resolveRelativePathWithinDir` as you suggested.

CalebJohn commented on packages/app-desktop/gui/NoteEditor/NoteBody/CodeMirror/utils/useExternalPlugins.ts:
```diff
@@ -0,0 +1,69 @@
+
+	function addInlineCss(cssStrings: string[], id: string) {
+		const element = document.createElement('style');
+		element.setAttribute('id', `${id}-inline`);
```
Thanks, I've changed this in both places.

CalebJohn commented on packages/app-desktop/gui/NoteEditor/NoteBody/CodeMirror/utils/useExternalPlugins.ts:
```diff
@@ -0,0 +1,69 @@
+		for (const contentScript of contentScripts) {
+			const mod = contentScript.module;
+			for (const addon of mod.addons) {
+				require(`codemirror/addon/${addon}`);
```
The error was clear, but it comes with a huge traceback that isn't helpful. I decided to update the error to be more clear and link to https://codemirror.net/doc/manual.html#addons.

CalebJohn commented on packages/app-desktop/gui/NoteEditor/NoteBody/CodeMirror/utils/useExternalPlugins.ts:
```diff
@@ -0,0 +1,69 @@
+
+		const contentScripts = contentScriptsToCodeMirrorPlugin(plugins);
+
+		for (const contentScript of contentScripts) {
```
Thanks! I decided to log an error instead of a warning. In this case it's a critical error because the plugin won't work at all.

CalebJohn commented:
@laurent22 sorry for the long diff. Most of the additions here are from the new contentScript example that I added under `codemirror_content_script`

These commits respond to your previous comments and also changed the `addons` field to `codeMirrorAssets` because I extended it to allow requiring anything in the codemirror directory which will be necessary to allow plugins to add support for more highlighting modes (becuase we only include the top ~20).

laurent22 commented on packages/app-desktop/gui/NoteEditor/NoteBody/CodeMirror/utils/useExternalPlugins.ts:
```diff
@@ -0,0 +1,91 @@
+				}
+			} catch (error) {
+				reg.logger().error(error.toString());
+				continue;
```
Just a small thing but I think this `continue` could be removed?

CalebJohn replied:
I put that there in case the next person editing the file placed code outside the try/catch block, in that case the `continue` would be necessary to ensure correct behaviour. In hindsight we shouldn't place anything outside the try/catch block so this won't be necessary (and won't be helpful).

laurent22 reviewed (commented):
That looks good, thanks for the update, and the detailed description on the demo plugin is very useful. I couldn't find if the demo file was linked from somewhere in the doc. If not already done, please could you link to it from `ContentScriptType` enum definition? (in plugins/api/types.ts)

CalebJohn commented:
@laurent22 thanks for that, I had added a link in `plugins/api/JoplinPlugins.ts` but I didn't think to add a reference to `types.ts`. I've added some documentation to it now. I didn't build the docs, I hope thats okay.

laurent22 commented:
Great, thanks for the update @CalebJohn, it's ready to merge now.

CalebJohn commented:
Thanks @laurent22 !
